src/test/java/org/junitpioneer/jupiter/CartesianProductTestExtensionTests.java

```diff
@@ -0,0 +1,185 @@
+/*
+ * Copyright 2015-2020 the original author or authors.
+ *
+ * All rights reserved. This program and the accompanying materials are
+ * made available under the terms of the Eclipse Public License v2.0 which
+ * accompanies this distribution and is available at
+ *
+ * http://www.eclipse.org/legal/epl-v20.html
+ */
+
+package org.junitpioneer.jupiter;
+
+import static org.junitpioneer.testkit.assertion.PioneerAssert.assertThat;
+
+import java.util.concurrent.TimeUnit;
+
+import org.assertj.core.api.Assertions;
+import org.junit.jupiter.api.DisplayName;
+import org.junit.jupiter.api.Nested;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.TestInfo;
+import org.junitpioneer.testkit.ExecutionResults;
+import org.junitpioneer.testkit.PioneerTestKit;
+
+public class CartesianProductTestExtensionTests {
+
+	@Nested
+	class StandardBehaviouralTests {
+
+		@CartesianProductTest(value = { "0", "1" })
+		@DisplayName("Does nothing if there are no parameters")
+		void empty() {
+		}
+
+		@CartesianProductTest(value = { "0", "1", "2" })
+		@DisplayName("Runs for each parameter once for single parameter")
+		void singleParameter(String param) {
+			int value = Integer.parseInt(param);
+			Assertions.assertThat(value).isBetween(0, 2);
+		}
+
+		@CartesianProductTest({ "0", "1" })
+		void threeBits(String a, String b, String c) {
+			int value = Integer.parseUnsignedInt(a + b + c, 2);
+			Assertions.assertThat(value).isBetween(0b000, 0b111);
+		}
+
+		@CartesianProductTest
+		@DisplayName("S ⨯ T ⨯ U")
+		void nFold(String string, Class<?> type, TimeUnit unit, TestInfo info) {
+			Assertions.assertThat(string).endsWith("a");
+			Assertions.assertThat(type).isInterface();
+			Assertions.assertThat(unit.name()).endsWith("S");
+			Assertions.assertThat(info.getTags()).isEmpty();
+		}
+
+		@CartesianProductTest(factory = "supplyValues")
+		@DisplayName("Can have an explicit factory method over the implicit one")
+		void explicitFactory(String string, TimeUnit unit) {
+			Assertions.assertThat(string).isIn("War", "Peace");
+			Assertions.assertThat(unit.name()).endsWith("S");
+		}
+
+	}
+
+	static CartesianProductTest.Sets nFold() {
+		return new CartesianProductTest.Sets()
+				.add("Alpha", "Omega")
+				.add(Runnable.class, Comparable.class, TestInfo.class)
+				.add(TimeUnit.DAYS, TimeUnit.HOURS);
+	}
+
+	static CartesianProductTest.Sets supplyValues() {
+		return new CartesianProductTest.Sets().add("War", "Peace").add(TimeUnit.SECONDS, TimeUnit.DAYS);
```

Comment:
Linebreaks for each `.add` like in `nFold`.

Reply:
Spotless. 🤷 
Would you prefer I did `@formatter:off`?

Reply:
To be honest: Yes, Think it's more readable then and consistent like in `nFold`

Reply:
Alright. Fair enough.